sw/device/tests/hmac_enc_irq_test.c

```diff
@@ -0,0 +1,147 @@
+
+  switch (interrupt_id) {
+    // Digest verification is done in a separate test, thus the digest is
+    // discarded.
+    case kTopEarlgreyPlicIrqIdHmacHmacDone:
+      hmac_irq = kDifHmacIrqHmacDone;
+      dif_hmac_digest_t dummy_digest;
+      dif_result_t res = dif_hmac_finish(&hmac, &dummy_digest);
+      CHECK(res == kDifOk, "HMAC finish failure = %d", res);
+      hmac_done_irq = true;
+      break;
+    case kTopEarlgreyPlicIrqIdHmacFifoEmpty:
+      hmac_irq = kDifHmacIrqFifoEmpty;
+      CHECK_DIF_OK(dif_hmac_irq_acknowledge(&hmac, hmac_irq));
+      hmac_empty_irq = true;
+      break;
+    default:
+      LOG_FATAL("ISR is not implemented!");
+      test_status_set(kTestStatusFailed);
+  }
+}
+
+/**
+ * External ISR.
+ *
+ * Handles all peripheral interrupts on Ibex. PLIC asserts an external interrupt
+ * line to the CPU, which results in a call to this OTTF ISR. This ISR
+ * overrides the default OTTF implementation.
+ */
+void ottf_external_isr(void) {
+  // Claim the IRQ by reading the Ibex specific CC register.
+  dif_rv_plic_irq_id_t interrupt_id;
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&plic, kPlicTarget, &interrupt_id));
+
+  // Check if the interrupted peripheral is HMAC.
+  top_earlgrey_plic_peripheral_t peripheral_id =
+      top_earlgrey_plic_interrupt_for_peripheral[interrupt_id];
+  CHECK(peripheral_id == kTopEarlgreyPlicPeripheralHmac, "Not an HMAC IRQ!");
+
+  handle_hmac_isr(interrupt_id);
+
+  // Complete the IRQ by writing the IRQ source to the Ibex specific CC
+  // register.
+  CHECK_DIF_OK(dif_rv_plic_irq_complete(&plic, kPlicTarget, interrupt_id));
+}
+
+/**
+ * Enables interrupts required by this test.
+ */
+static void enable_irqs(void) {
+  mmio_region_t base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR);
+  CHECK_DIF_OK(dif_rv_plic_init(base_addr, &plic));
+
+  // Enable interrupts in HMAC IP.
+  CHECK_DIF_OK(
+      dif_hmac_irq_set_enabled(&hmac, kDifHmacIrqHmacDone, kDifToggleEnabled));
+  CHECK_DIF_OK(
+      dif_hmac_irq_set_enabled(&hmac, kDifHmacIrqFifoEmpty, kDifToggleEnabled));
+
+  // Enable interrupts in PLIC.
+  rv_plic_testutils_irq_range_enable(&plic, kTopEarlgreyPlicTargetIbex0,
+                                     kTopEarlgreyPlicIrqIdHmacHmacDone,
+                                     kTopEarlgreyPlicIrqIdHmacFifoEmpty);
+
+  // Enable interrupts in Ibex.
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+}
+
+bool test_main() {
+  LOG_INFO("Running HMAC DIF IRQ test...");
+
+  mmio_region_t base_addr = mmio_region_from_addr(TOP_EARLGREY_HMAC_BASE_ADDR);
+  CHECK_DIF_OK(dif_hmac_init(base_addr, &hmac));
+
+  hmac_done_irq = false;
+  hmac_empty_irq = false;
+  enable_irqs();
+
+  // The purpose of this test is to ensure that HMAC empty and done interrupts
+  // assert when the the conditions are met. Digest is not verified by this
+  // test, which means that a "dummy" data filled with zeros will suffice.
+  uint32_t sent_bytes;
+  char data[4] = {0};
```

Comment:
I would prefer to use any data different from 0x00 and 0xFF as it can be confused when debugging the traces.

Reply:
Well, it doesn't matter for this test, as the data is irrelevant, we simply trying to fire an interrupt. But I can change it to A5A5.

Reply:
Done.